<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Marker simulation — markerSimfb • fbnet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Marker simulation — markerSimfb"><meta property="og:description" content="Simulates marker genotypes conditional on the pedigree structure, affection
statuses and disease model."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fbnet</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/MarsicoFL/fbnet.git/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Marker simulation</h1>
    <small class="dont-index">Source: <a href="https://github.com/MarsicoFL/fbnet.git/blob/HEAD/R/markerSimfb.R" class="external-link"><code>R/markerSimfb.R</code></a></small>
    <div class="hidden name"><code>markerSimfb.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Simulates marker genotypes conditional on the pedigree structure, affection
statuses and disease model.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">markerSimfb</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  N <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  available <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alleles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  afreq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  partialmarker <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  loop_breakers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  eliminate <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>a <code>linkdat</code> object</p></dd>


<dt>N</dt>
<dd><p>a positive integer: the number of markers to be simulated</p></dd>


<dt>available</dt>
<dd><p>a vector containing IDs of the available individuals, i.e.
those whose genotypes should be simulated. By default, all individuals are
included.</p></dd>


<dt>alleles</dt>
<dd><p>a vector containing the alleles for the marker to be
simulation. If a single integer is given, this is interpreted as the number
of alleles, and the actual alleles as <code>1:alleles</code>. Must be NULL if
<code>partialmarker</code> is non-NULL.</p></dd>


<dt>afreq</dt>
<dd><p>a vector of length 2 containing the population frequencies for
the marker alleles. Must be NULL if <code>partialmarker</code> is non-NULL.</p></dd>


<dt>partialmarker</dt>
<dd><p>Either NULL (resulting in unconditional simulation), a
marker object (on which the simulation should be conditioned) or the index
of an existing marker of <code>x</code>.</p></dd>


<dt>loop_breakers</dt>
<dd><p>a numeric containing IDs of individuals to be used as
loop breakers. Relevant only if the pedigree has loops, and only if
<code>partialmarker</code> is non-NULL. See <code>breakLoopsfb</code>.</p></dd>


<dt>eliminate</dt>
<dd><p>A non-negative integer, indicating the number of iterations
in the internal genotype-compatibility algorithm. Positive values can save
time if <code>partialmarker</code> is non-NULL and the number of alleles is
large.</p></dd>


<dt>seed</dt>
<dd><p>NULL, or a numeric seed for the random number generator.</p></dd>


<dt>verbose</dt>
<dd><p>a logical.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a <code>linkdat</code> object equal to <code>x</code> except its
<code>markerdata</code> entry, which consists of the <code>N</code> simulated markers.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This implements (with various time savers) the algorithm used in SLINK of the
LINKAGE/FASTLINK suite. If <code>partialmarker</code> is NULL, genotypes are
simulated by simple gene dropping, using <code>simpleSim</code>.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>G. M. Lathrop, J.-M. Lalouel, C. Julier, and J. Ott,
<em>Strategies for Multilocus Analysis in Humans</em>, PNAS 81(1984), pp.
3443-3446.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>simpleSim</code>, <code>linkage.power</code></p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ariel Chernomoretz, Franco Marsico.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

